src/app/content/search/hooks/index.ts

```diff
@@ -52,45 +50,40 @@ export const receiveSearchHook: ActionHookBody<typeof receiveSearchResults> = (s
     : getFirstResult(book, payload);
 
   if (
-    // selectedResult may equal to null if api did not return any results
-    !selectedResult
+    selectedResult &&
     // We are clearing selected result when requesting a new search so in the theory this should never happen
-    || isEqual(select.selectedResult(state), selectedResult)
+    (isEqual(select.selectedResult(state), selectedResult)
     // selectedResult bookId data is different than current book id
-    || book.id !== getIndexData(selectedResult.result.index).bookId
+    || book.id !== getIndexData(selectedResult.result.index).bookId)
   ) {
     return;
   }
 
-  services.dispatch(selectSearchResult(selectedResult));
-
-  const targetPageId = selectedResult.result.source.pageId;
-  const targetPage = assertDefined(
-    findArchiveTreeNodeById(book.tree, targetPageId),
-    'search result pointed to page that wasn\'t in book'
-  );
-
-  // currentPage may by undefined when user started a search from the 404 page
-  const page = currentPage || targetPage;
-
-  const navigation = {
-    params: getBookPageUrlAndParams(book, targetPage).params,
-    route: content,
-    state : {
-      bookUid: book.id,
-      bookVersion: book.version,
-      pageUid: stripIdVersion(targetPage.id),
-    },
-  };
-
-  const action = stripIdVersion(page.id) === stripIdVersion(targetPage.id) ? replace : push;
-  const search = queryString.stringify({
-    query,
-    target: JSON.stringify({ type: 'search', index: selectedResult.highlight }),
-  });
-  const hash = selectedResult.result.source.elementId;
+  if (selectedResult) {
+    services.dispatch(selectSearchResult(selectedResult));
+  }
 
-  services.dispatch(action(navigation, { hash, search }));
+  const targetPageId = selectedResult?.result.source.pageId || currentPage?.id;
+
+  const action = (targetPageId && stripIdVersion(targetPageId)) === (currentPage && stripIdVersion(currentPage.id))
+    ? replace : push;
+
+  const options = selectedResult ?
+    {
+      hash: selectedResult.result.source.elementId,
+      search: queryString.stringify({
+        query,
+        target: JSON.stringify({ type: 'search', index: selectedResult.highlight }),
+      }),
+    } :
+    {
+      search: queryString.stringify({ query }),
+    };
+
+  const navigation = targetPageId ? createRouteMatchOptions(targetPageId, book) : selectNavigation.match(state);
```

Comment:
this method name is `createRouteMatchOptions` but it is being assigned to something called `navigation`, and another thing called `options` is defined above, something is not right here